### PR TITLE
fix(motion_utils): add doxygen comments for utility functions in trajectory header

### LINKED ISSUE #2803: Add brief functions description doxygen comments for motion_utils/trajectory.hpp
### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I've agreed with the maintainers that I can plan this task.

### Description

Utility functions in trajectory.hpp under motion_utils is one of the files that is widely used in autoware planning. 
However, a lot of these functions are missing brief description and other doxygen commenting 

### Purpose

To have a clearer description and code documentation for these utility functions that will help better re-use them.

### Possible approaches

Add doxygen comments before each function in motion_utlis/trajectory.hpp

### Definition of done

- [ ] Add doxygen comments before all functions in trajectory.hpp

### PR DESCRIPTION
**Pull request by ahmeddesokyebrahim**
Signed-off-by: AhmedEbrahim <ahmed.a.d.ebrahim@gmail.com>

## Description
closes #2803 

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: common/motion_utils/include/motion_utils/trajectory/trajectory.hpp [hunk 1]
```diff
@@ -1238,6 +1450,15 @@ double calcSignedArcLength(
   return signed_length_on_traj - signed_length_src_offset;
 }
 
+/**
+ * @brief calculate length of 2D distance between two points, specificed by start piont index in
+ * points contrainer and end point and its segment index in points container
```
Comment (takayuki5168) [inline common/motion_utils/include/motion_utils/trajectory/trajectory.hpp:1455]: In some parts, the indent is wrong.

Comment (ahmeddesokyebrahim) [reply common/motion_utils/include/motion_utils/trajectory/trajectory.hpp]: That is fixed in my new commit.
Thanks Mr. @takayuki5168 for reviewing it.

**Review by brkay54 — COMMENTED** [review]
Thank you for your work! I just added some minor changes.

File: common/motion_utils/include/motion_utils/trajectory/trajectory.hpp [hunk 2]
```diff
@@ -340,10 +417,13 @@ boost::optional<size_t> findNearestSegmentIndex(
 }
 
 /**
- * @brief calculate lateral offset from p_target (length from p_target to trajectory)
- *        If seg_idx point is after that nearest point, length is negative
+ * @brief calculate lateral offset from p_target (length from p_target to trajectory) using given
+ * segment index. If seg_idx point is after that nearest point, length is negative. Segment is
```
[the remark this answers is not in the document]
Comment (ahmeddesokyebrahim) [reply common/motion_utils/include/motion_utils/trajectory/trajectory.hpp]: Fixed

File: common/motion_utils/include/motion_utils/trajectory/trajectory.hpp [hunk 3]
```diff
@@ -414,7 +505,12 @@ double calcLateralOffset(
 }
 
 /**
- * @brief calcSignedArcLength from index to index
+ * @brief calculate length of 2D distance between two points, specified by start and end points
+ * indicies through points container.
+ * @param points points of trajectory, path, ...
+ * @param src_idx index of start point
+ * @param dst_idx index of end point
+ * @return length of distance between two points
```
[the remark this answers is not in the document]
Comment (ahmeddesokyebrahim) [reply common/motion_utils/include/motion_utils/trajectory/trajectory.hpp]: Fixed in this place and other place with similar comment.

File: common/motion_utils/include/motion_utils/trajectory/trajectory.hpp [hunk 4]
```diff
@@ -383,6 +463,17 @@ double calcLateralOffset(
   return cross_vec(2) / segment_vec.norm();
 }
 
+/**
+ * @brief calculate lateral offset from p_target (length from p_target to trajectory).
+ * The function gets the nearest segment index between the points of trajectory and the given target
+ * point, then uses that segment index to calculate lateral offset. If seg_idx point is after that
+ * nearest point, length is negative. Segment is straight path between two continuous points of
```
[the remark this answers is not in the document]
Comment (ahmeddesokyebrahim) [reply common/motion_utils/include/motion_utils/trajectory/trajectory.hpp]: Fixed

**Review by brkay54 — APPROVED** [review]
LGTM!

**Review by takayuki5168 — APPROVED** [review]
LGTM

**Review by takayuki5168 — CHANGES_REQUESTED** [review]
test

File: common/motion_utils/include/motion_utils/trajectory/trajectory.hpp [hunk 5]
```diff
@@ -438,40 +534,49 @@ double calcSignedArcLength(const T & points, const size_t src_idx, const size_t
...
   std::vector<double> partial_dist;
   partial_dist.reserve(dst_idx - src_idx);
 
   double dist_sum = 0.0;
   partial_dist.push_back(dist_sum);
   for (size_t i = src_idx; i < dst_idx - 1; ++i) {
-    dist_sum += tier4_autoware_utils::calcDistance2d(points.at(i), points.at(i + 1));
     partial_dist.push_back(dist_sum);
+    dist_sum += tier4_autoware_utils::calcDistance2d(points.at(i), points.at(i + 1));
   }
```
Comment (takayuki5168) [inline common/motion_utils/include/motion_utils/trajectory/trajectory.hpp:567]: Is this change okay? This common function is already used in the behavior_path_planner.

Comment (ahmeddesokyebrahim) [reply common/motion_utils/include/motion_utils/trajectory/trajectory.hpp:567]: Thanks for your comments Mr. @takayuki5168 .
I am just trying to understand the change request and its context : 
- Is it only deleting the line before puch_back and making it after it ?
  - if yes. that means the first 2 slots of the vector will be the same, as we already puch_back before the loop. Is it right ?
- Is it OK to do a code change in this PR in the time it was originally open for doxygen comments ?

Comment (ahmeddesokyebrahim) [reply common/motion_utils/include/motion_utils/trajectory/trajectory.hpp]: Mr. @takayuki5168 : 
I have understood the problem, it was due to a merge conflict mistake in one of previous commit and I did not realize it.

Please review again let me know if you have further findings.

Thanks again

**Review by takayuki5168 — APPROVED** [review]
LGTM

**Pull request merged**